### PR TITLE
Generate default config for each request

### PR DESCRIPTION
**Pull request by MartinForReal**
related issue: https://github.com/nats-io/nats.go/issues/308
Signed-off-by: Fan Shang Xiang <fanshangxiang@gmail.com>

### CODE REVIEW

File: stan.go [hunk 1]
```diff
@@ -303,16 +319,37 @@ type ack struct {
...
+	if o.ConnectTimeout == 0 {
+		o.ConnectTimeout = DefaultConnectWait
 	}
+	if o.DiscoverPrefix == "" {
+		o.DiscoverPrefix = DefaultDiscoverPrefix
+	}
+	if o.MaxPubAcksInflight == 0 {
+		o.MaxPubAcksInflight = DefaultMaxInflight
+	}
+	if o.PingIterval == 0 {
```
Comment (wallyqs) [inline stan.go]: Interesting, this option has a typo...

Comment (MartinForReal) [reply stan.go]: should I rename this option to PingInterval?

Comment (wallyqs) [reply stan.go]: Since it has been public for a while probably have to deprecate the option and replace with `PingInterval`

Comment (MartinForReal) [reply stan.go]: deprecate option as suggested

**Comment by kozlovic** [comment]
@MartinForReal Thank you for the contribution. There are things in this PR that we could use, namely fixing the PingInterval option. However, we don't want to go to Options.Connect(). We had that in NATS and moved to the Connect() with passing options.

If you feel that some option setters are missing, then we should add them. Of course, there is always the possibility for a user to create their own function to set the option - if it is missing from the library and they need it urgently.

The other thing that we could use from this PR is deprecating DefaultOptions and introducing the GetDefaultOptions function.

**Comment by MartinForReal** [comment]
hi @kozlovic, I got a scenario where I need to get the option from command arguments. And it is very convenient to bind variable address with argument options. So I don't need to define another struct to store config values. I guess one of the reasons why Options.Connect() is abandoned is because there is no validation process when the config is set. And I think the validation process can be added in Options.Connect(). How do you think of the idea?

**Comment by kozlovic** [comment]
@MartinForReal no, the choice to move away from `Options.Connect()` and to a `Connect(...Option)` was not because of the lack of validation. This could be added as you did in the PR.
It was chosen to favor the model where we pass function setters for setting the options, and will be odd to go back to old model now.
Plus, the Options.Connect() has a flaw whereby unless the connection created there does create its own copy of options (or clone them) then if the connection references options and those options are changed concurrently by the user, that could cause panics/race conditions.

**Comment by MartinForReal** [comment]
Thank you for clarifying it.

**Pull request closed without merging**